### PR TITLE
Ignore close errors in read-only files.

### PR DESCRIPTION
**Pull request by christianpoveda**
this fixes https://github.com/rust-lang/miri/issues/999

r? @RalfJung

### CODE REVIEW

File: src/shims/fs.rs [hunk 1]
```diff
@@ -148,6 +152,13 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         let fd = this.read_scalar(fd_op)?.to_i32()?;
 
         if let Some(handle) = this.machine.file_handler.handles.remove(&fd) {
+            // We avoid syncing the file if it was opened in read-only mode.
```
Comment (RalfJung) [inline src/shims/fs.rs]: Maybe add a link to some discussion explaining why.

Comment (christianpoveda) [reply src/shims/fs.rs]: I added a brief explanation and a link to #999

File: src/shims/fs.rs [hunk 2]
```diff
@@ -148,13 +152,24 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
         if let Some(handle) = this.machine.file_handler.handles.remove(&fd) {
-            // `File::sync_all` does the checks that are done when closing a file. We do this to
-            // to handle possible errors correctly.
-            let result = this.try_unwrap_io_result(handle.file.sync_all().map(|_| 0i32));
-            // Now we actually close the file.
-            drop(handle);
-            // And return the result.
-            result
+            // We sync the file if it was opened in a mode different than read-only.
+            if !handle.read_only {
```
Comment (RalfJung) [inline src/shims/fs.rs]: Unless there are good arguments for the opposite, it's better to put the positive case first (`if handle.read_only`). That avoids a mental double-negation when considering the `else` branch.

File: src/shims/fs.rs [hunk 3]
```diff
@@ -148,13 +152,24 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+                // And return the result.
+                result
+            } else {
+                // We drop the file, this closes it but ignores any errors produced when closing
+                // it. This is done because `File::sync_call` cannot be done over files like
+                // `/dev/urandom`. Check
+                // https://github.com/rust-lang/miri/issues/999#issuecomment-568920439 for a deeper
+                // discussion.
+                drop(handle);
+                return Ok(0);
```
Comment (RalfJung) [inline src/shims/fs.rs]: Why do you early-return here? Wouldn't just `Ok(0)` (without `return`) also work?

File: src/shims/fs.rs (hunk 2 above)
Comment (christianpoveda) [reply src/shims/fs.rs]: I did it this way to keep the "usual" behavior first and then be able to explain why this exception is required.

File: src/shims/fs.rs (hunk 3 above)
Comment (christianpoveda) [reply src/shims/fs.rs]: Yes,

File: src/shims/fs.rs (hunk 2 above)
Comment (RalfJung) [reply src/shims/fs.rs]: Why is read-write more "usual" than read-only?

Comment (christianpoveda) [reply src/shims/fs.rs]: huh that's a good question. I don't think that's the case. Taking that into account, should I change the field inside the handle to read-write then?

Comment (RalfJung) [reply src/shims/fs.rs]: Whatever you prefer, as long as you avoid the negation here. ;)

Comment (christianpoveda) [reply src/shims/fs.rs]: I changed it to avoid the double negation.

**Comment by RalfJung** [comment]
Looking good, thanks!

@bors r+

**Pull request merged**